grml-debootstrap

```diff
@@ -1507,96 +1497,96 @@ grub_install() {
 
-  if [ -n "$ARM_EFI_TARGET" ]; then
-    einfo "Installing Grub as bootloader into EFI."
-
-    # shellcheck disable=SC2086
-    clean_chroot "$MNTPOINT" DEBIAN_FRONTEND=$DEBIAN_FRONTEND apt-get -y --no-install-recommends install $DPKG_OPTIONS grub-efi-arm64 grub-efi-arm64-signed
-
-    clean_chroot "${MNTPOINT}" grub-install --target=arm64-efi --efi-directory=/boot/efi --bootloader-id=debian --recheck --no-nvram --removable
   # Has chroot-script installed GRUB to MBR using grub-install (successfully), already?
   # chroot-script skips installation for unset ${GRUB}
-  elif [[ -z "${GRUB}" ]] || ! dd if="${GRUB}" bs=512 count=1 2>/dev/null | cat -v | grep -Fq GRUB; then
+  if [[ -z "${GRUB}" ]] || ! dd if="${GRUB}" bs=512 count=1 2>/dev/null | cat -v | grep -Fq GRUB; then
     einfo "Installing Grub as bootloader."
 
-    if ! clean_chroot "${MNTPOINT}" dpkg --list grub-pc 2>/dev/null | grep -q '^ii' ; then
-      echo "Notice: grub-pc package not present yet, installing it therefore."
-      # shellcheck disable=SC2086
-      clean_chroot "$MNTPOINT" DEBIAN_FRONTEND=$DEBIAN_FRONTEND apt-get -y --no-install-recommends install $DPKG_OPTIONS grub-pc
+    local grub_pc_package_name=''
+
+    # ARM64 uses EFI-only, and doesn't need a GRUB package. For i386, we only
+    # install the BIOS bootloader if we're not using EFI (since grub-pc and
+    # grub-efi-ia32 can't be co-installed). For amd64, we install grub-cloud
+    # which will pull in grub-pc-bin, so we don't need to explicitly install a
+    # BIOS bootloader for it when using EFI.
+    if [ -z "$VMEFI" ]; then
+      grub_pc_package_name='grub-pc'
     fi
 
-    mkdir -p "${MNTPOINT}/boot/grub"
-    if ! [ -d "${MNTPOINT}"/usr/lib/grub/i386-pc/ ] ; then
-      eerror "Error: grub not installed inside Virtual Machine. Can not install bootloader."
-      bailout 1
+    if [ -n "${grub_pc_package_name}" ]; then
+      if ! clean_chroot "${MNTPOINT}" dpkg --list "${grub_pc_package_name}" 2>/dev/null | grep -q '^ii' ; then
+        echo "Notice: ${grub_pc_package_name} package not present yet, installing it therefore."
+        clean_chroot "$MNTPOINT" DEBIAN_FRONTEND=$DEBIAN_FRONTEND apt-get -y --no-install-recommends install $DPKG_OPTIONS "${grub_pc_package_name}"
+      fi
     fi
-    cp -a "${MNTPOINT}"/usr/lib/grub/i386-pc "${MNTPOINT}/boot/grub/"
 
-    if [ -n "$VMEFI" ]; then
+    mkdir -p "${MNTPOINT}/boot/grub"
 
+    if [ -n "$VMEFI" ]; then
       mkdir -p "${MNTPOINT}"/boot/efi
       mount -t vfat "${EFI_TARGET}" "${MNTPOINT}"/boot/efi
 
+      efi_id="$(
+        if [ -f "${MNTPOINT}"/etc/default/grub ]; then
+          source "${MNTPOINT}"/etc/default/grub
+        fi
+        for file in "${MNTPOINT}"/etc/default/grub.d/*.cfg; do
+          if [ -f "${file}" ]; then
+            source "${file}"
+          fi
+        done
+        if [ "${GRUB_DISTRIBUTOR}" != "" ]; then
+          echo "${GRUB_DISTRIBUTOR}"
+        else
+          echo 'debian'
+        fi
+      )"
+
```

Comment:
This contradicts the:

  `The --efi-id option and GRUB_DISTRIBUTOR should be removed, yes.`

nor? What's you use case exactly for this? BTW, if at all the `debian` should be `Debian` to match Debian's defaults?

Reply:
The `--efi-id` option and `GRUB_DISTRIBUTOR` writing stuff is gone. I left this here because it's a far more flexible approach than hardcoding - in the event Debian changes their `GRUB_DISTRIBUTOR` in the future for whatever reason, this will automatically catch it and use that EFI ID without an explicit change needing to be made. It's generally good to avoid hardcoding, so I thought it was worthwhile to keep.

> BTW, if at all the debian should be Debian to match Debian's defaults?

Did this change in Trixie+? Under Bookworm, `/boot/efi/EFI/debian` is the bootloader directory, not `/boot/efi/EFI/Debian`.

Reply:
> Did this change in Trixie+? Under Bookworm, /boot/efi/EFI/debian is the bootloader directory, not /boot/efi/EFI/Debian

I imagine the case-change is irrelevant on FAT filesystems. I do see a change in how GRUB_DISTRIBUTOR is derived between bookworm and trixie, so maybe it is an artifact of that.

Reply:
Ah I see. What I was referring to was this line in `/etc/default/grub` , which looks on Debian unstable like this:

```
% grep DIST /etc/default/grub
GRUB_DISTRIBUTOR=`( . /etc/os-release; echo ${NAME:-Debian} ) 2>/dev/null || echo Debian`
```

But even on bullseye AKA v11 it's similar:

```
 % grep DISTR /etc/default/grub 
GRUB_DISTRIBUTOR=`lsb_release -i -s 2> /dev/null || echo Debian`
```

Reply:
Aha, that makes sense. I guess since we're parsing `GRUB_DISTRIBUTOR` from the existing config (not writing our own new config) anyway, using the default shown there of `Debian` would be acceptable. It probably won't have any effect, but it seems more correct now that I see this.

Reply:
Case is now changed so that if all else fails, `Debian` is used.